Implement unbuffered channels

MichaelKim20 opened the pull request:
I Implemented the default channel semantic of Go, unbuffered channels that provide direct hands off semantics.

There are two kinds of fiber(thread). One reads, and one writes.

**Unbuffered channel**

When the write-fiber(thread) is working, 
it sends the data to the read-fiber(thread) if it already exists.
Otherwise, the write-fiber(thread) waits for the read-fiber(thread) to come in.

When read-fiber(thread) works, 
it receives data from the write-fiber(thread) if it already exists.
Otherwise, the read-fiber(thread) waits for the write-fiber(thread) to come in.

The issue is #41 Implement unbuffered channels

Geod24 commented on source/geod24/concurrency.d:
```diff
@@ -898,6 +924,51 @@ public class InfoThread : Thread
+ * and may be instantiated and used, but is not a necessary part of the
+ * default functioning of this module.
+ */
+class ThreadScheduler
```
?

MichaelKim20 replied:
I made a mistake, I corrected it.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1602,7 +1602,32 @@ private:
-    This is similar to go buffered channels
+    This is similar to go channels.
+
+    If you set the size of the queue to 0 in creator,
```
```suggestion
    If you set the size of the queue to 0 in the constructor,
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1643,14 +1692,56 @@ public class Channel (T)
 
-        this.queue.insertBack(msg);
-        return true;
+        scope scheduler = thisScheduler;
```
I would add a comment here:

```
// queue is full or channel is unbuffered
```

AndrejMitrovic replied:
```suggestion
        auto scheduler = thisScheduler;
```

Just a note here: `scope` has no meaning when you assign existing values to variables. It only has meaning when you instantiate a new object. For example:

```D
scope obj = new Class();  // destroyed at end of scope
```

This is different to the following:

```D
auto obj = new Class();
scope var = obj;  // *not* destroyed at end of scope
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1643,14 +1692,56 @@ public class Channel (T)
-        this.queue.insertBack(msg);
-        return true;
+        scope scheduler = thisScheduler;
+        if (Fiber.getThis() && scheduler)
```
```suggestion
        if (Fiber.getThis() && scheduler !is null)
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1643,14 +1692,56 @@ public class Channel (T)
-        return true;
+        scope scheduler = thisScheduler;
+        if (Fiber.getThis() && scheduler)
+        {
```
There's duplication in these two blocks. So I suggest rewriting it:

```D
void waitForReader (Mutex mutex, Condition condition)
{
    ChannelContext!T new_context;
    new_context.msg_ptr = null;
    new_context.msg = msg;
    new_context.mutex = null;
    new_context.condition = condition;
    this.sendq.insertBack(new_context);
    this.mutex.unlock();
    new_context.wait();
}

scope scheduler = thisScheduler;
if (Fiber.getThis() && scheduler)
{
    waitForReader(null, scheduler.newCondition(null));
    return true;
}
else
{
    auto mutex = new Mutex();
    waitForReader(mutex, new Condition(mutex));
    return true;
}
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1665,12 +1756,58 @@ public class Channel (T)
+        if (this.closed)
+        {
+            this.mutex.unlock();
+            assert(0, "Channel is closed.");
```
Very nice. This is how assertions should be used, to catch programmer errors.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1665,12 +1756,58 @@ public class Channel (T)
+            assert(0, "Channel is closed.");
+        }
+
         T msg;
```
I would prefer if `T msg;` was declared inside of each `if` block separately. Because it will not be re-used. So:

```D
if (!this.sendq.empty)
{
    T msg;
    ...
}

if (!this.queue.empty)
{
    T msg = this.queue.front;
    ...
}
```

And so on..

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1665,12 +1756,58 @@ public class Channel (T)
+        scope scheduler = thisScheduler;
+        if (Fiber.getThis() && scheduler)
+        {
+            ChannelContext!T new_context;
```
These two `if` / `else` blocks again have similar code, so I suggest writing a helper function just like the one I suggested in the `send` routine.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1665,12 +1756,58 @@ public class Channel (T)
+        if (!this.queue.empty)
+        {
+            msg = this.queue.front;
+            this.queue.removeFront();
```
```suggestion
            this.queue.removeFront();
            assert(this.mcount > 0, "mcount was zero!");
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1691,19 +1828,32 @@ public class Channel (T)
         if (!this.queue.empty)
         {
             *(msg) = this.queue.front;
             this.queue.removeFront();
```
There is a missing `this.mcount--;` decrement here. The tests need to be improved because they didn't catch this error.

```suggestion
            this.queue.removeFront();
            assert(this.mcount > 0, "mcount was zero!");
            this.mcount--;
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1734,38 +1884,155 @@ public class Channel (T)
 
```
The method documentation needs to be updated. It doesn't say what closing the channel actually does. And it looks like it notifies all waiting receivers/senders.

AndrejMitrovic commented:
It's a bit strange that https://github.com/Geod24/localrest/pull/44 introduces a `Channel` structure (also included as a commit in this PR), but then this PR #45 immediately changes the code introduced in #44 to allow the channel to be buffered or un-buffered.

This kind of atomic split of the Channel implementation would only make sense if the two PRs were introduced with a long time between them, where the first implementation was used by LocalRest/Agora for a while.

So I suggest merging the two commits into one, which says something like "Introduced (un)buffered Channels".

MichaelKim20 commented:
> It's a bit strange that #44 introduces a `Channel` structure (also included as a commit in this PR), but then this PR #45 immediately changes the code introduced in #44 to allow the channel to be buffered or un-buffered.
> 
> This kind of atomic split of the Channel implementation would only make sense if the two PRs were introduced with a long time between them, where the first implementation was used by LocalRest/Agora for a while.
> 
> So I suggest merging the two commits into one, which says something like "Introduced (un)buffered Channels".

@AndrejMitrovic 's opinion is good.
By the way, the two PRs belong to different issues. And I think I'd like to clarify the technical differences between the two and how they're implemented.
@Geod24 What do you think about this?

MichaelKim20 commented:
@Geod24  @AndrejMitrovic  Ready for review

MichaelKim20 commented:
@AndrejMitrovic Thank you for your kind review.

MichaelKim20 commented:
I merged the two commits into one,
PR #44 and PR #45,
A message of the new commit is "Implement (un)buffered Channel"
@Geod24 @AndrejMitrovic  Ready for review.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1603,3 +1604,449 @@ private:
+        this.mutex.lock();
+        scope (exit) this.mutex.unlock();
+
+        this.closed = true;
```
Does it make sense to call `close` on a channel multiple times? Because if not then there should be another assert here:

```suggestion
        assert(!this.closed, "Channel was already closed!");
        this.closed = true;
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1603,3 +1604,449 @@ private:
+
+        this.closed = true;
+
+        while (true)
```
```suggestion
        while (!this.recvq.empty)
```

and then you can get rid of the `if / break` below.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1603,3 +1604,449 @@ private:
+        this.queue.clear();
+        this.mcount = 0;
+
+        while (true)
```
```suggestion
        while (!this.sendq.empty)
```

same comment as above.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1603,3 +1604,449 @@ private:
+    /// This is a message. Used in put
+    public T  msg;
+
+    /// This is a message point. Used in get
```
```suggestion
    /// This is a message pointer. Used in get
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -824,6 +824,9 @@ struct ThreadInfo
     Tid owner;
 
+    /// Storage of information required for scheduling, message passing, etc.
+    public Object[string] objects;
```
I am not sure why this is here. This can just be `FiberScheduler scheduler;`, and `thisScheduler` could return it. Why is it a map, and why does it store `Object` instead of `FiberScheduler`?

It is not used to store any other types of objects in this PR.

MichaelKim20 replied:
All Commitments can be found at PR #25.
It will be used later. What should I do in this case?

https://github.com/Geod24/localrest/pull/25/commits/297220c01b9530b3b12513422d1b05b7f1781222#diff-b2ce8db5a10ee88e89a581259145e669R115-R147

AndrejMitrovic replied:
Alright, so you want to store multiple types into `ThreadInfo`. But why not just store:

```
struct ThreadInfo
{
    FiberScheduler scheduler;
    MessageChannel channel;
}
```

And then add more things as it's necessary? It doesn't have to use a hashmap and be dynamic.

AndrejMitrovic replied:
Ok Michael explained it to me IRL, so the issue is that `ThreadInfo` is in `geod24.concurrency`, but the real type we want to store there is `Channel!Type`, and we don't know the type yet.

There might be a better way to do this, but I'll have to think about it. In the meantime, we could go with Michael's approach.

MichaelKim20 replied:
Thank you for your review.

I made the following corrections.
struct ThreadInfo
{
    FiberScheduler scheduler;
}
If I need an Object later, I will add a full explanation there.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+    t1.start();
+}
+
+/// Min Thread -> [ channel1 ] -> Thread1 -> [ channel2 ] -> Min Thread
```
```suggestion
/// Parent thread and child thread communicating with 2 channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+}
+
+
+/// Min Thread -> [ channel1 ] -> Fiber1 in Thread1 -> [ channel2 ] -> Min Thread
```
```suggestion
/// Parent thread and fiber of child thread communicating with 2 channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+    assert(channel2.receive() == 4);
+}
+
+/// Fiber in Min Thread -> [ channel1 ] -> Fiber1 in Thread1 -> [ channel2 ] -> Fiber in Min Thread
```
```suggestion
/// Parent thread's fiber and child thread communicating with 2 channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+    assert(result == 4);
+}
+
+/// Fiber1 -> [ channel2 ] -> Fiber2 -> [ channel1 ] -> Fiber1
```
```suggestion
/// 2 fibers in a child thread communicating with each other using 2 channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+}
+
+
+/// Fiber1 in Thread1 -> [ channel2 ] -> Fiber2 in Thread2 -> [ channel1 ] -> Fiber1 in Thread1
```
```suggestion
/// Fiber in parent thread communicating with fiber in child thread using 2 channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+}
+
+
+/// Thread1 -> [ channel2 ] -> Thread2 -> [ channel1 ] -> Thread1
```
```suggestion
/// Two child threads communicating with each other with 2 channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+}
+
+
+/// Thread1 -> [ channel2 ] -> Fiber1 in Thread 2 -> [ channel1 ] -> Thread1
```
```suggestion
/// One child thread communicating with another child thread's fiber using 2 channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+    Mutex mutex = new Mutex;
+    Condition condition = new Condition(mutex);
+
+    // Thread1 - It'll be tangled.
```
The word we use in concurrency is `blocked`, not `tangled`.

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+    }
+    assert(result == 0);
+
+    // Thread2 - Unravel a tangle
```
```suggestion
    // Thread2 - Unblock a thread
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+    assert(result == 2);
+
+    result = 0;
+    // Thread3 - It'll not be tangled, because queue size is 1
```
```suggestion
    // Thread3 - It'll not be blocked, because queue size is 1
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+}
+
+
+// If the queue size is 0, it will block when it is sent and received on the same thread.
```
```suggestion
// Test blocking behavior on threads when using unbuffered channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+}
+
+
+// If the queue size is 0, it will block when it is sent and received on the same fiber.
```
```suggestion
// Test blocking behavior on fibers when using unbuffered channels
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+            cond.wait(1000.msecs);
+            assert(result == 2);
+
+            //  Fiber3 - It'll not be tangled, because queue size is 1
```
```suggestion
            //  Fiber3 - It'll not be blocked, because queue size is 1
```

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -2040,3 +2040,331 @@ unittest
+            assert(!cond.wait(1000.msecs));
+            assert(result == 0);
+
+            //  Fiber2 - Unravel a tangle
```
```suggestion
            //  Fiber2 - Unblock a fiber
```

AndrejMitrovic commented:
The tests so far look good, however they are incomplete.

There are missing tests for:
- Return value of `send`. They are ignored in all the test-cases.
- There also need to be tests that cause `send` to return false.
- Test calling close()
- Test calling send after calling close()
- Test calling `tryReceive` after calling close(). Unlike `receive` which will `assert` if trying to read on a closed channel, `tryReceive` will return false. It needs a test.
- isClosed() test

MichaelKim20 commented:
@AndrejMitrovic Thank you for your review

MichaelKim20 commented:
> The tests so far look good, however they are incomplete.
> 
> There are missing tests for:
> 
> * Return value of `send`. They are ignored in all the test-cases.
> * There also need to be tests that cause `send` to return false.
> * Test calling close()
> * Test calling send after calling close()
> * Test calling `tryReceive` after calling close(). Unlike `receive` which will `assert` if trying to read on a closed channel, `tryReceive` will return false. It needs a test.
> * isClosed() test

@AndrejMitrovic I checked 'std.concuracy.send'.
There was no return value, and I confirmed using 'assert' inside.
Many people will find it inconvenient to use 'Channel.send' because they have to check the return value every time they use it.
So I changed the return of 'Channel.send' to void.
And I added the test code before and after closing the channel.

MichaelKim20 commented:
@AndrejMitrovic @Geod24 Ready for review

AndrejMitrovic commented on source/geod24/concurrency.d:
```diff
@@ -1572,3 +1599,800 @@ private:
+        Params:
+            msg = message to send
+
+        Return:
```
Need to remove `Return` part.

AndrejMitrovic reviewed (approved):
Just one small thing left.

AndrejMitrovic commented:
> @AndrejMitrovic I checked 'std.concurency.send'.
There was no return value, and I confirmed using 'assert' inside.
Many people will find it inconvenient to use 'Channel.send' because they have to check the return value every time they use it.
So I changed the return of 'Channel.send' to void.
And I added the test code before and after closing the channel.

There is one issue: `std.concurrency.send` will not fail to send a request. In the Channel implementation, sending to a closed channel will fail, right? So, you would use code like this:

```D
if (!channel.isClosed())
    channel.send(...);
```

But there is a problem here. There is a race condition. After `isClosed` returns, the mutex is released, and then another thread can call `channel.close`. Meanwhile the main thread will think the channel is not closed, it will call `send()`, and it will `assert(0)` in the body.

So I think we should keep the return value of `send`.

MichaelKim20 commented:
> > @AndrejMitrovic I checked 'std.concurency.send'.
> > There was no return value, and I confirmed using 'assert' inside.
> > Many people will find it inconvenient to use 'Channel.send' because they have to check the return value every time they use it.
> > So I changed the return of 'Channel.send' to void.
> > And I added the test code before and after closing the channel.
> 
> There is one issue: `std.concurrency.send` will not fail to send a request. In the Channel implementation, sending to a closed channel will fail, right? So, you would use code like this:
> 
> ```d
> if (!channel.isClosed())
>     channel.send(...);
> ```
> 
> But there is a problem here. There is a race condition. After `isClosed` returns, the mutex is released, and then another thread can call `channel.close`. Meanwhile the main thread will think the channel is not closed, it will call `send()`, and it will `assert(0)` in the body.
> 
> So I think we should keep the return value of `send`.

I'll think more about this. Thank you.

AndrejMitrovic commented:
@MukeunKim actually your code is ok. I've had a look at how Go channels are used, and it's the writer thread that ever calls `close` on a channel.

In the future, it would be good to add the thread owner ID to the Channel, so if the wrong thread calls `.close()` on a channel then it would assert.

But for now, your code is good!

MichaelKim20 commented:
@Geod24 Ready for review

MichaelKim20 committed:
Implement (un)buffered Channel

MichaelKim20 committed:
Add `FiberScheduler` in `ThreadInfo`, add function thisScheduler.

MichaelKim20 committed:
Add test codes for message passing between thread and thread, thread and fiber, fiber and fiber
